crates/wasmtime/src/engine.rs

```diff
@@ -79,6 +128,53 @@ impl Engine {
         Arc::ptr_eq(&a.inner, &b.inner)
     }
 
+    pub(crate) fn register_module_signatures(
+        &self,
+        signatures: &PrimaryMap<SignatureIndex, WasmFuncType>,
+        trampolines: impl Iterator<Item = (SignatureIndex, VMTrampoline)>,
+    ) -> (SharedSignatures, TrampolineMap) {
+        self.inner
+            .signatures
+            .write()
+            .unwrap()
+            .register_module(signatures, trampolines)
+    }
+
+    pub(crate) fn register_signature(&self, ty: &WasmFuncType) -> VMSharedSignatureIndex {
+        self.inner.signatures.write().unwrap().register(ty)
+    }
+
+    pub(crate) fn unregister_signatures(
```

Comment:
Having read a bit more, I guess the `SharedSignatures` structure sort of is this RAII type, but for a whole collection of signatures.

I do think that this function should still be marked `unsafe` though, and have a comment laying out why, as discussed up above.

Reply:
One final thing: since nothing can guarantee that we aren't currently "borrowing" an index out of the `SharedSignatures` by the time we drop it and unregister its indices (because the indices aren't references and/or types with a lifetime and they are used by JIT code and such) I think that actually creating a `SharedSignatures` should be `unsafe` and have the contract that "by the time you drop this, you promise you aren't 'borrowing' any indices from within it anymore".

Reply:
Ok I've reworked this such that `Engine`, `Module`, and `Store` all internally use a `SignatureCollection` that has a reference on the signature registry.  This is an RAII type that will unregister the signatures on drop.

This cleans up the code quite a bit and makes it less error prone.  I'll push up the change soon.